chapter_recurrent-modern/seq2seq.ipynb

```diff
@@ -0,0 +1,1066 @@
+   },
+   "source": [
+    "Now we can create and train an RNN encoder-decoder model\n",
+    "for sequence to sequence learning on the machine translation dataset.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int embedSize = 32,\n",
+    "        numHiddens = 32,\n",
+    "        numLayers = 2,\n",
+    "        batchSize = 64,\n",
+    "        numSteps = 10,\n",
+    "        numEpochs = 300;\n",
+    "float dropout = 0.1f, lr = 0.005f;\n",
+    "Device device = Functions.tryGpu(0);\n",
+    "\n",
+    "Pair<ArrayDataset, Pair<Vocab, Vocab>> dataNMT =\n",
+    "        NMT.loadDataNMT(batchSize, numSteps, 600, manager);\n",
+    "ArrayDataset dataset = dataNMT.getKey();\n",
+    "Vocab srcVocab = dataNMT.getValue().getKey();\n",
+    "Vocab tgtVocab = dataNMT.getValue().getValue();\n",
+    "\n",
+    "encoder = new Seq2SeqEncoder(srcVocab.length(), embedSize, numHiddens, numLayers, dropout);\n",
+    "decoder = new Seq2SeqDecoder(tgtVocab.length(), embedSize, numHiddens, numLayers, dropout);\n",
+    "\n",
+    "EncoderDecoder net = new EncoderDecoder(encoder, decoder);\n",
+    "trainSeq2Seq(net, dataset, lr, numEpochs, tgtVocab, device);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "ZhqzmGf9mMWF",
+    "origin_pos": 35
+   },
+   "source": [
+    "## Prediction\n",
+    "\n",
+    "To predict the output sequence\n",
+    "token by token,\n",
+    "at each decoder time step\n",
+    "the predicted token from the previous\n",
+    "time step is fed into the decoder as an input.\n",
+    "Similar to training,\n",
+    "at the initial time step\n",
+    "the beginning-of-sequence (\"&lt;bos&gt;\") token\n",
+    "is fed into the decoder.\n",
+    "This prediction process\n",
+    "is illustrated in :numref:`fig_seq2seq_predict`.\n",
+    "When the end-of-sequence (\"&lt;eos&gt;\") token is predicted,\n",
+    "the prediction of the output sequence is complete.\n",
+    "\n",
+    "\n",
+    "![Predicting the output sequence token by token using an RNN encoder-decoder.](https://github.com/d2l-ai/d2l-en-colab/blob/master/img/seq2seq-predict.svg?raw=1)\n",
+    ":label:`fig_seq2seq_predict`\n",
+    "\n",
+    "We will introduce different\n",
+    "strategies for sequence generation in\n",
+    ":numref:`sec_beam-search`.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "/* Predict for sequence to sequence. */\n",
+    "public static Pair<String, ArrayList<NDArray>> predictSeq2Seq(\n",
+    "        EncoderDecoder net,\n",
+    "        String srcSentence,\n",
+    "        Vocab srcVocab,\n",
+    "        Vocab tgtVocab,\n",
+    "        int numSteps,\n",
+    "        Device device,\n",
+    "        boolean saveAttentionWeights)\n",
+    "        throws IOException, TranslateException {\n",
+    "    Integer[] srcTokens =\n",
```

Comment:
```suggestion
    "    int[] srcTokens =\n",
```

Reply:
Originally I have Integer because in the previous section we use getIdxs and that returns a list. The first thing that truncatePad does it convert it to int though. Is this okay?

Reply:
The Integer is object while int is primitive type.  To reduce GC pressure, we tend to use int as much as possible. But if there is constraint to use Integer in prev section, it's fine